src/brevitas/core/function_wrapper/clamp.py

```diff
@@ -73,3 +81,203 @@ def __init__(self, min_val: float) -> None:
+        self.saturating = saturating
+
+        # inf without NaN not possible
+        if self.inf_values is None and self.nan_values is None:
+            self.max_val_impl = StatelessBuffer(
+                max_float(
+                    self.exponent_bit_width(), self.mantissa_bit_width(), self.exponent_bias()))
+        elif self.inf_values is not None and self.nan_values is not None:
+            # we have values for NaN and inf, so initiate MaxValInfNaN
+            self.max_val_impl = MaxFloatInfNaN(
+                exponent_bit_width=self.exponent_bit_width,
+                mantissa_bit_width=self.mantissa_bit_width,
+                exponent_bias=self.exponent_bias,
+                nan_values=self.nan_values,
+                inf_values=self.inf_values,
+                saturating=self.saturating)
+        elif self.inf_values is None and self.nan_values is not None:
+            # we only have values for NaN, so initiate MaxValNaN
+            self.max_val_impl = MaxFloatNaN(
+                exponent_bit_width=self.exponent_bit_width,
+                mantissa_bit_width=self.mantissa_bit_width,
+                exponent_bias=self.exponent_bias,
+                nan_values=self.nan_values,
+                saturating=self.saturating)
+        else:
+            # no NaN values but inf values
+            raise RuntimeError('Minifloat Error: inf value cannot exist without NaN value.')
+
+        self.clamp_impl = CaseClamp(inf_values=self.inf_values, saturating=self.saturating)
+
+    @brevitas.jit.script_method
+    def forward(self, inp: Tensor):
+        # get max value for the minifloat config
+        max_value = self.max_val_impl()
+        return self.clamp_impl(inp, max_value)
+
+
+class MaxFloatInfNaN(brevitas.jit.ScriptModule):
+
+    def __init__(
+            self,
+            exponent_bit_width: Tensor,
+            mantissa_bit_width: Tensor,
+            exponent_bias: Tensor,
+            nan_values: Tuple[str],
+            inf_values: Tuple[str],
+            saturating: bool = False) -> None:
+        super(MaxFloatInfNaN, self).__init__()
+        self.exponent_bit_width = exponent_bit_width
+        self.mantissa_bit_width = mantissa_bit_width
+        self.exponent_bias = exponent_bias
+
+        self.inf_values = inf_values
+        self.nan_values = nan_values
+
+        self.saturating = saturating
+
+        # check that NaN/inf values are all mantissa_bit_width long
+        if any(map(lambda x: len(x) > mantissa_bit_width, self.nan_values + self.inf_values)):
+            raise RuntimeError('NaN/inf codes need to be the same length as the mantissa.')
+
+    @brevitas.jit.script_method
+    def forward(self):
+        # idea: take inf and nan values, select the smallest, set max_value to smallest_val - 1
+        min_special_case = min(map(lambda x: int(x, 2), self.nan_values + self.inf_values))
+        max_value_mantissa = min_special_case - 1
+
+        if max_value_mantissa < 0:
```

Comment:
This chunk of code is repeated below, so it must go somewhere else. Maybe a function?